File: Day_12_Hill_Climbing_Algorithm/Python/main.py
import numpy as np
from numpy.typing import NDArray
import heapq
from typing import List, Tuple


# class representing 2d elevation grid
class Grid:
    def __init__(self, input_lines: List[str]):
        # length of grid
        self.l = len(input_lines)
        # width of grid
        self.w = len(input_lines[0])
        # 2d elevation grid
        self.elevation = np.zeros((self.l, self.w), dtype=np.int8)
        # starting point
        self.start = (0, 0)
        # ending point
        self.end = (0, 0)
        # set of points with zero elevation
        self.zero_elevation = set()

        # iterate over grid rows
        for i in range(self.l):
            # iterate over grid columns
            for j in range(self.w):
                # get elevation value from character
                v, is_start, is_end = self.get_elevation(input_lines[i][j])

                # add to zero elevation set if zero
                if v == 0:
                    self.zero_elevation.add((i, j))

                # set elevation value
                self.elevation[i, j] = v

                # set start or end points
                if is_start:
                    self.start = (i, j)
                elif is_end:
                    self.end = (i, j)

    def get_elevation(self, char: str) -> Tuple[int, bool, bool]:
        """
        Get elevation value and whether it's a start or end point, given a character.
        """
        # whether it's a start point
        is_start = False
        # whether it's an end point
        is_end = False

        # a-z elevation character
        elevation_char = char

        # start point
        if elevation_char == 'S':
            elevation_char = 'a'
            is_start = True
        # end point
        elif elevation_char == 'E':
            elevation_char = 'z'
            is_end = True

        # convert character to elevation value
        elevation = ord(elevation_char) - ord('a')

        return elevation, is_start, is_end

    def get_neighbours(self, i: int, j: int, visited: NDArray[np.bool_], reverse: bool = False) -> List[Tuple[int, int]]:
        """
        Get neighbouring indices. This returns only indices of neighbours that can be
        travelled to from the current point. If reverse = True, this returns indices of
        neighbours that can be travelled from to the current point.
        """
        candidates = [
            # top
            (i - 1, j),
            # bottom
            (i + 1, j),
            # left
            (i, j - 1),
            # right
            (i, j + 1),
        ]
        # set of neighbouring indices
        neighbours = set()

        for _i, _j in candidates:
            # not a valid neighbour if indices out of range
            if _i < 0 or _i >= self.l or _j < 0 or _j >= self.w:
                continue

            # not a valid neighbour if already visited
            if visited[_i, _j]:
                continue

            # not a valid neighbour if higher difference in elevation than a single step
            if not reverse and self.elevation[_i, _j] - self.elevation[i, j] > 1:
                continue

            # not a valid neighbour if lower difference in elevation than a single step
            if reverse and self.elevation[i, j] - self.elevation[_i, _j] > 1:
                continue

            # add indices to set of neighbours
            neighbours.add((_i, _j))

        return neighbours

    def multiple_sources_shortest_paths(self, source: Tuple[int, int], reverse: bool) -> NDArray[np.int64]:
        """
        Perform Dikstra's algorithm to obtain shortest paths from given source to
        multiple sources.
        """
        # boolean array representing whether indices have already been visited
        visited = np.zeros((self.l, self.w), dtype=np.bool_)
        visited[source] = True

        # shortest distances from each pair of indices, initialized to inf
        distances = np.full((self.l, self.w), np.iinfo(np.int64).max, dtype=np.int64)

        # priority queue with distance and indices
        # each item in queue is a 3-element tuple
        # first element is distance, so priority queue is always ordered on distance
        queue = [(0,) + source]

        while len(queue) > 0:
            # pop queue item
            d, i, j = heapq.heappop(queue)
            # relaxation
            if d < distances[i, j]:
                distances[i, j] = d

            # push neighbours onto queue
            for _i, _j in self.get_neighbours(i, j, visited, reverse=reverse):
                visited[_i, _j] = True
                heapq.heappush(queue, (d + 1, _i, _j))

        return distances


if __name__ == '__main__':
    # file with elevation data
    file_path = '../elevation.txt'
    # open and read file
    with open(file_path) as f:
        file_contents = f.read()

    # create grid by parsing file contents
    grid = Grid(file_contents.splitlines())
    # compute shortest paths to end point
    distances = grid.multiple_sources_shortest_paths(grid.end, True)
    # shortest path from start to end
    print(distances[grid.start])
    # shortest path from any zero elevation point to end
    print(min([distances[i, j] for (i, j) in grid.zero_elevation]))

File: Day_12_Hill_Climbing_Algorithm/Go/main.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"math"
	"os"
	"unicode/utf8"
)

// struct representing row and column coordinates
type Coordinates struct {
	// row index
	I int
	// column index
	J int
}

// priority queue item
type QueueItem struct {
	// distance from source to item
	Distance int64
	// index of item in sorted queue
	Index int
	// coordinates of item
	Coor Coordinates
}

// priority queue for breadth first traversal
type Queue []*QueueItem

// get length of priority queue
func (queue Queue) Len() int {
	return len(queue)
}

// comparision priority queue items by index
func (queue Queue) Less(i, j int) bool {
	return queue[i].Distance < queue[j].Distance
}

// swap priority queue items by index
func (queue Queue) Swap(i, j int) {
	queue[i], queue[j] = queue[j], queue[i]
	queue[i].Index = i
	queue[j].Index = j
}

// push new item into priority queue
func (queue *Queue) Push(x any) {
	n := len(*queue)
	queueItem := x.(*QueueItem)
	queueItem.Index = n
	*queue = append(*queue, queueItem)
}

// pop smallest item from priority queue
// in the context of this problem, it is the coordinates with shortest distance
func (queue *Queue) Pop() any {
	oldQueue := *queue
	n := len(oldQueue)
	queueItem := oldQueue[n-1]
	oldQueue[n-1] = nil
	queueItem.Index = -1
	*queue = oldQueue[0 : n-1]

	return queueItem
}

// struct representing elevation grid
type Grid struct {
	// 2d slice representing elevation
	Elevation [][]int
	// length of grid
	L int
	// width of grid
	W int
	// starting coordinates
	Start Coordinates
	// ending coordinates
	End Coordinates
	// slice of coordinates with the lowest elevation
	Lowest []Coordinates
}

// get elevation at given coordinates
func (grid *Grid) E(c Coordinates) int {
	return grid.Elevation[c.I][c.J]
}

// get neighbours of given coordinates
// neighbours are the set of coordinates from which the traverser can move to the given coordinates
func (grid *Grid) GetNeighbours(c Coordinates, visited map[Coordinates]bool) []Coordinates {
	// possible candidates for neighbours
	candidates := []Coordinates{
		// top
		{I: c.I - 1, J: c.J},
		// bottom
		{I: c.I + 1, J: c.J},
		// left
		{I: c.I, J: c.J - 1},
		// right
		{I: c.I, J: c.J + 1},
	}

	// slice of valid neighbours
	neighbours := []Coordinates{}

	for _, candidate := range candidates {
		// not a valid neighbour if indices out of range
		if candidate.I < 0 || candidate.I >= grid.L || candidate.J < 0 || candidate.J >= grid.W {
			continue
		}

		// not a valid neighbour if already visited
		_, ok := visited[candidate]
		if ok {
			continue
		}

		// not a valid neighbour if neighbour's elevation is higher than one level above that of current coordinates
		if grid.E(c)-grid.E(candidate) > 1 {
			continue
		}

		// add to slice of neighbours
		neighbours = append(neighbours, candidate)
	}

	return neighbours
}

// perform Dijkstra's algorithm to compute distances from source to all destinations
// in the current context, the source is the ending coordinates
func (grid *Grid) Dijkstra(source Coordinates) map[Coordinates]int64 {
	// map of visited coordinates
	visited := make(map[Coordinates]bool)
	// set source as visited
	visited[source] = true

	// map of current distances to coordinates
	distances := make(map[Coordinates]int64)
	// priority queue to store coordinates
	queue := make(Queue, 1)
	// add source coordinates to priority queue
	queue[0] = &QueueItem{
		Coor:     source,
		Distance: 0,
		Index:    0,
	}
	heap.Init(&queue)

	// loop until queue is empty
	for queue.Len() > 0 {
		// get next queue item
		queueItem := heap.Pop(&queue).(*QueueItem)
		// if queue item distance is shorter than currently stored distance, update stored distance
		currDistance, ok := distances[queueItem.Coor]
		if !ok || queueItem.Distance < currDistance {
			distances[queueItem.Coor] = queueItem.Distance
		}

		// get neighbouring coordinates
		neighbours := grid.GetNeighbours(queueItem.Coor, visited)
		for _, neighbour := range neighbours {
			// set neighbour as visisted
			visited[neighbour] = true
			// push neighbour to priority queue
			heap.Push(
				&queue,
				&QueueItem{
					Distance: queueItem.Distance + 1,
					Coor:     neighbour,
				},
			)
		}
	}

	return distances
}

// get elevation from character, and whether it's a starting/ending point
func GetElevation(c rune) (int, bool, bool) {
	isStart := false
	isEnd := false

	// 'S' represents starting point and has the same elevation as 'a'
	if c == 'S' {
		isStart = true
		c = 'a'
		// 'E' represents ending point and has the same elevation as 'z'
	} else if c == 'E' {
		isEnd = true
		c = 'z'
	}

	return int(c) - int('a'), isStart, isEnd
}

// get line-by-line file scanner
func GetFileScanner(filePath string) (*bufio.Scanner, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s, %s", filePath, err.Error())
	}

	scanner := bufio.NewScanner(f)

	return scanner, nil
}

func main() {
	// file with elevation data
	filePath := "../elevation.txt"

	scanner, err := GetFileScanner(filePath)
	if err != nil {
		fmt.Println("failed to get file scanner", err.Error())
		return
	}

	// create grid object
	grid := Grid{}
	// initialize 2d slice of elevation
	grid.Elevation = make([][]int, 0)
	// initialize starting, ending, and lowest coordinates
	grid.Start = Coordinates{}
	grid.End = Coordinates{}
	grid.Lowest = make([]Coordinates, 0)

	// row index
	i := 0
	// scan file line-by-line
	for scanner.Scan() {
		rowText := scanner.Text()
		rowLen := utf8.RuneCountInString(rowText)
		// declare slice of row
		rowElevation := make([]int, rowLen)
		// set grid width
		grid.W = rowLen

		for j, c := range rowText {
			// get and set elevation value
			elevation, isStart, isEnd := GetElevation(c)
			rowElevation[j] = elevation

			// set starting/endpoint points
			if isStart {
				grid.Start.I = i
				grid.Start.J = j
			} else if isEnd {
				grid.End.I = i
				grid.End.J = j
			}

			// add to lowest elevations slice
			if elevation == 0 {
				grid.Lowest = append(grid.Lowest, Coordinates{I: i, J: j})
			}
		}

		// add row of elevations to grid
		grid.Elevation = append(grid.Elevation, rowElevation)
		// increment row index
		i++
	}

	// set grid length
	grid.L = i
	// compute distances using Dijkstra's algorithm
	distances := grid.Dijkstra(grid.End)
	// print distance from starting point
	fmt.Println(distances[grid.Start])

	// iterate over computed distances and find shortest distance among lowest elevation coordinates
	shortestDistance := math.MaxInt64
	for _, start := range grid.Lowest {
		distance, ok := distances[start]
		if ok && distance < int64(shortestDistance) {
			shortestDistance = int(distance)
		}
	}
	fmt.Println(shortestDistance)
}
